database/scrapers/polinka_stations.ts

```diff
@@ -0,0 +1,113 @@
+import * as fs from "node:fs/promises";
+import { Readable } from "node:stream";
+
+import { BaseScraperModule, TaskHandle } from "#commands/db_scrape";
+import { ExternalDigitalGuideMode } from "#enums/digital_guide_mode";
+import Polinka from "#models/polinka_station";
+import FilesService from "#services/files_service";
+
+const assetsPath = "https://admin.topwr.solvro.pl/assets/";
+const filesMetaPath = (id: string) =>
+  `https://admin.topwr.solvro.pl/files/${id}?fields=filename_disk`;
+
+interface SourceResponse<T> {
+  data: T[];
+}
+
+interface PolinkaStationDraft {
+  name: string;
+  latitude: number;
+  longitude: number;
+  address_line1: string;
+  address_line2: string;
+  photoKey: string | null;
+  externalDigitalGuideIdOrURL: string;
+  externalDigitalGuideMode: ExternalDigitalGuideMode | null;
+}
+
+function isValidDataResponse<T>(
+  response: unknown,
+): response is SourceResponse<T> {
+  return (
+    typeof response === "object" &&
+    response !== null &&
+    "data" in response &&
+    Array.isArray(response.data)
+  );
+}
+
+const isValidPolinkaStationData = (
+  data: unknown,
+): data is SourceResponse<PolinkaStationDraft> =>
+  isValidDataResponse<PolinkaStationDraft>(data);
+
+export default class PolinkaStationScraper extends BaseScraperModule {
+  static name = "Polinka";
+  static description =
+    'Scrapes polinkas stations from local file: "./assets/polinkas.json"';
+  static taskTitle = "Scrape Polinka";
+
+  async run(task: TaskHandle) {
+    task.update("starting reading polinkas file...");
+
+    const polinkaStationsData = await fs
+      .readFile("./assets/polinka_stations.json", { encoding: "utf-8" })
+      .then(JSON.parse)
+      .then((data) => {
+        if (!isValidPolinkaStationData(data)) {
+          throw new Error(`
+            Invalid JSON structure in ./assets/polinka_stations.json,
+            expected type of Response<PolinkaStationDraft>
+            `);
+        }
+        return data;
+      });
+
+    for (const polinkaStation of polinkaStationsData.data) {
+      polinkaStation.photoKey = await this.semaphore.runTask(() =>
+        this.uploadCoverAndGetKey(polinkaStation),
+      );
+    }
```

Comment:
this doesnt need to be in a semaphore if you're awaiting the promise in the loop.
you'd need to use the semaphore if you used a function like `Promise.all`.